ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -1038,22 +1038,29 @@ withConnectionManager ConnectionManagerArguments {
                       shouldTrace = absConnState /= TerminatedSt
                       isTerminating = absConnState' == TerminatingSt
 
-                  -- 'handleError' might be either a handshake negotiation
-                  -- a protocol failure (an IO exception, a timeout or
-                  -- codec failure).  In the first case we should not reset
-                  -- the connection as this is not a protocol error.
-                  writeTVar connVar connState'
-
                   updated <-
-                    modifyTMVarPure
+                    modifyTMVarSTM
                       stateVar
                       ( \state ->
                         case Map.lookup peerAddr state of
-                          Nothing -> (state, False)
+                          Nothing -> return (state, False)
                           Just mutableConnState'  ->
                             if mutableConnState' == mutableConnState
-                              then (Map.delete peerAddr state , True)
-                              else (state                     , False)
+                              then do
+                                -- 'handleError' might be either a handshake negotiation
+                                -- a protocol failure (an IO exception, a timeout or
+                                -- codec failure).  In the first case we should not reset
+                                -- the connection as this is not a protocol error.
+                                --
+                                -- If we are deleting the connState from the the state
+                                -- then connState' can be TerminatingSt in which case we
+                                -- are going to transition TerminatingSt -> TerminatedSt.
+                                -- Otherwise, Connection Manager cleanup will take care of
+                                -- tracing accordingly.
```

Comment:
I am a bit confused by this explanation.  Do you actually mean that `connState` can be `TerminatingSt`?  Certainly if `connState'` is `TerminatingSt` then we're doing `someState -> TerminationSt` transition, not `TerminatingSt -> TerminatedSt`.

Reply:
Could you create an issue for this fixed simulation bug (use the right labels, check the other such issues), please describe the test failure there.

Reply:
Yes, it can be `TerminatingSt`, so we have to move the `writeTVar` inside this block to avoid tracing an out of place transition. I will add an issue for this fixed simulation bug

Reply:
fixed

Reply:
done!